client/my-sites/email/email-forwarding/email-forwarding-add-new-compact-list.jsx

```diff
@@ -0,0 +1,178 @@
+
+class EmailForwardingAddNewCompactList extends Component {
+	static propTypes = {
+		emailForwards: PropTypes.array,
+		emailForwardingLimit: PropTypes.number,
+		onConfirmEmailForwarding: PropTypes.func.isRequired,
+		onSuccessRedirectDestination: PropTypes.string,
+		selectedDomainName: PropTypes.string.isRequired,
+	};
+
+	constructor( props ) {
+		super( props );
+		this.state = {
+			emailForwards: [ { destination: '', mailbox: '', isValid: false } ],
+			isRedirecting: false,
+			newForwardAdded: false,
+		};
+	}
+
+	hasValidEmailForwards() {
+		return ! this.state.emailForwards?.some( ( forward ) => ! forward.isValid );
+	}
+
+	addNewEmailForwardWithAnalytics = ( domainName, mailbox, destination ) => {
+		withAnalytics(
+			composeAnalytics(
+				this.props.onConfirmEmailForwarding(),
+				this.props.addEmailForward( domainName, mailbox, destination )
+			)
+		);
+	};
+
+	addNewEmailForwardsClick = ( event ) => {
+		const { selectedSiteSlug, selectedDomainName } = this.props;
+
+		event.preventDefault();
+
+		if ( this.props.isAddingEmailForward ) {
+			return;
+		}
+
+		this.state.emailForwards?.map( ( forward ) => {
+			const { mailbox, destination } = forward;
+
+			this.addNewEmailForwardWithAnalytics(
+				selectedDomainName,
+				mailbox,
+				destination,
+				selectedSiteSlug
+			);
+		} );
+		this.state.newForwardAdded = true;
+	};
+
+	onForwardAdd = () => {
+		this.setState( {
+			emailForwards: [ ...this.state.emailForwards, { destination: '', mailbox: '' } ],
+		} );
+	};
+
+	renderActionsButtons() {
+		const { translate } = this.props;
+		return (
+			<div className="email-forwarding-add-new-compact-list__actions">
+				<Button
+					primary
+					onClick={ this.addNewEmailForwardsClick }
+					disabled={ ! this.hasValidEmailForwards() }
+				>
+					{ translate( 'Add' ) }
+				</Button>
+			</div>
+		);
+	}
+
+	onRemoveEmailForward = ( index ) => {
+		const [ emailForwards ] = this.state;
+		emailForwards.splice( index, 1 );
+		this.setState( { emailForwards } );
+	};
+
+	onUpdateEmailForward = ( index, name, value ) => {
+		// eslint-disable-next-line prefer-const
+		let emailForwards = this.state.emailForwards;
+		emailForwards[ index ][ name ] = value;
+
+		const validEmailForward = validateAllFields( emailForwards[ index ] );
+		emailForwards[ index ].isValid =
+			validEmailForward.mailbox.length === 0 && validEmailForward.destination.length === 0;
+
+		this.setState( { emailForwards } );
+	};
+
+	componentDidUpdate() {
+		const {
+			emailForwardSuccess,
+			isRequestingDomains,
+			onSuccessRedirectDestination,
+			siteId,
+		} = this.props;
+
+		const { isRedirecting, newForwardAdded } = this.state;
+
+		if (
+			emailForwardSuccess &&
+			onSuccessRedirectDestination &&
+			! isRedirecting &&
+			newForwardAdded
+		) {
+			if ( ! isRequestingDomains && ! isRedirecting ) {
+				this.props.fetchSiteDomains( siteId );
+			}
+			page( onSuccessRedirectDestination );
+			this.state.isRedirecting = true;
+		}
+	}
```

Comment:
All this guards are to prevent the following cases:

- We have added a forward on a previous step, so emailForwardSuccess is already true, so in case we are in purchase page, we are redirected automatically to email/manage!
- We don't redirect twice (or more) while redirecting
- We don't fetch while fetching domains
- We don't redirect without having previously added a new forward (newForwardAdded)

PS: I couldn't add fetchSiteDomains in the deconstruction as it shadows import